lib/Canonicalize/MakeModelCastPass.cpp

```diff
@@ -132,37 +133,35 @@ MMCP::serializeTypesForModelCast(FunctionMetadataCache &Cache,
       SerializeTypeFor(Ret->getOperandUse(0));
 
   } else if (auto *SI = dyn_cast<StoreInst>(I)) {
-    auto &PtrOperandPtrType = TypeMap.at(SI->getPointerOperand());
-    auto &ValOperandType = TypeMap.at(SI->getValueOperand());
-
-    const model::Architecture::Values &Arch = Model.Architecture();
-    QualifiedType ValOperandPtrType = ValOperandType.getPointerTo(Arch);
-    if (PtrOperandPtrType != ValOperandPtrType) {
-      bool IsPtrOperandOfAggregateType = false;
-      QualifiedType PtrOperandType;
-
-      if (PtrOperandPtrType.isPointer()) {
-        PtrOperandType = dropPointer(PtrOperandPtrType);
-        const auto *Unqualified = PtrOperandType.UnqualifiedType().getConst();
-        IsPtrOperandOfAggregateType = isa<model::StructType>(Unqualified)
-                                      || isa<model::UnionType>(Unqualified);
-      }
+    auto &PtrOperandPtrType = *TypeMap.at(SI->getPointerOperand());
+    auto &ValOperandType = *TypeMap.at(SI->getValueOperand());
+
+    bool IsPointerToAnAggregate = false;
+    const model::UpcastableType *PointeeType = nullptr;
+    if (auto Pointer = PtrOperandPtrType.tryGetAsPointer()) {
+      PointeeType = &Pointer->PointeeType();
```

Comment:
> This can be assigned only if you don't early-exit with the following `if`

But wait, the check itself uses the `PointeeType`. And it's an extremely cheap pointer assignment.